pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,633 @@
+			case <-stopCh:
+				return
+			// TODO; Isn't this always true? Is this for dealing with time drift?
+			case <-time.After(time.Until(e.electionStart)): // elections are time ordered in the channel
+				err := c.runElection(ctx, e.leaderLeaseID)
+				utilruntime.HandleError(err)
+			}
+		}
+	}
+}
+
+func (c *Controller) reconcileLeaseCandidate(lease *v1alpha1.LeaseCandidate) error {
+	if lease.Spec.TargetLease == "" {
+		return nil
+	}
+	klog.Infof("reconcile found canLead label namespace=%q, name=%q: %q", lease.Namespace, lease.Name, lease.Spec.TargetLease)
+	leaderLeaseID := leaderLeaseID{lease.Namespace, lease.Spec.TargetLease}
+
+	// Ignore candidates that transitioned to Pending because reelection is already in progress
+	if lease.Annotations[AckAnnotationName] == "Pending" {
+		return nil
+	} else if lease.Annotations[AckAnnotationName] == "Ack" {
+		// If the candidate ack'd the reelection request, mark it as
+		// so the candidate make
+		status := c.getCT(leaderLeaseID)
+		if status != nil {
+			status.Ack(lease.Name)
+		}
+		return nil
+	}
+
+	klog.Infof("Adding lease %s", leaderLeaseID.String())
+	c.leaseQueue.Add(leaderLeaseID.String())
+	return nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	lease, err := c.leaseInformer.Lister().Leases(namespace).Get(name)
+
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+	if lease != nil {
+		if lease.Annotations[ElectedByAnnotationName] == controllerName {
+			klog.Infof("reconcile for lease namespace=%q, name=%q", lease.Namespace, lease.Name)
+		} else if isLeaseExpired(lease) {
+			clone := lease.DeepCopy()
+			if lease.Annotations[ElectedByAnnotationName] == controllerName && lease.Spec.HolderIdentity != nil && clone.Spec.RenewTime != nil && clone.Spec.LeaseDurationSeconds != nil && clone.Spec.AcquireTime != nil {
+				delete(clone.Annotations, EndOfTermAnnotationName)
+				clone.ObjectMeta.Annotations[ElectedByAnnotationName] = controllerName
+				clone.Spec.HolderIdentity = nil
+				clone.Spec.RenewTime = nil
+				clone.Spec.LeaseDurationSeconds = nil
+				clone.Spec.AcquireTime = nil
+				_, err := c.leaseClient.Leases(clone.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+	}
+
+	err = c.scheduleElection(leaderLeaseID{namespace: namespace, name: name})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Controller) scheduleElection(leaderLeaseID leaderLeaseID) error {
+	// TODO: Create additional options for skipping elections for leases that opt out of CLE
+	if leaderLeaseID.name == "leader-election-controller" {
+		return nil
+	}
+	klog.Infof("scheduleElection")
+	// TODO: add a set to track ongoing elections and avoid requesting an election if one has already been kicked off?
+	// This might not be needed..
+	c.electionCh <- election{
```

Comment:
why don't run the election process on the worker? the worker consumes from the workqueue that guarantees that the same key is only processed once